flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateStreamReader.java

```diff
@@ -0,0 +1,111 @@
+
+import org.apache.flink.runtime.checkpoint.channel.ChannelStateReader.ReadResult;
+import org.apache.flink.runtime.checkpoint.channel.RefCountingFSDataInputStream.RefCountingFSDataInputStreamFactory;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.BufferBuilder;
+import org.apache.flink.runtime.state.AbstractChannelStateHandle;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+
+import static org.apache.flink.runtime.checkpoint.channel.ChannelStateByteBuffer.wrap;
+import static org.apache.flink.runtime.checkpoint.channel.ChannelStateReader.ReadResult.HAS_MORE_DATA;
+import static org.apache.flink.runtime.checkpoint.channel.ChannelStateReader.ReadResult.NO_MORE_DATA;
+
+/**
+ * Reads the state of a single channel pointed by {@link org.apache.flink.runtime.state.AbstractChannelStateHandle AbstractChannelStateHandle}.
+ * Once all data is read, this class can't be used anymore.
+ * Uses {@link RefCountingFSDataInputStream} internally.
+ */
+@NotThreadSafe
+class ChannelStateStreamReader implements Closeable {
+
+	private final RefCountingFSDataInputStream stream;
+	private final ChannelStateDeserializer serializer;
+	private final Queue<Long> offsets;
+	private int remainingBytes = -1;
+	private boolean closed = false;
+
+	ChannelStateStreamReader(AbstractChannelStateHandle<?> handle, RefCountingFSDataInputStreamFactory streamFactory) {
+		this(streamFactory.getOrCreate(handle), handle.getOffsets(), streamFactory.getSerializer());
+	}
+
+	private ChannelStateStreamReader(RefCountingFSDataInputStream stream, List<Long> offsets, ChannelStateDeserializer serializer) {
+		this.stream = stream;
+		this.stream.incRef();
+		this.serializer = serializer;
+		this.offsets = new LinkedList<>(offsets);
+	}
+
+	ReadResult readInto(Buffer buffer) throws IOException {
+		return readInto(wrap(buffer));
+	}
+
+	ReadResult readInto(BufferBuilder bufferBuilder) throws IOException {
+		return readInto(wrap(bufferBuilder));
+	}
+
+	private ReadResult readInto(ChannelStateByteBuffer buffer) throws IOException {
+		Preconditions.checkState(!closed, "reader is closed");
+		readWhilePossible(buffer);
+		if (haveMoreData()) {
+			return HAS_MORE_DATA;
+		} else {
+			closed = true;
+			stream.decRef();
+			return NO_MORE_DATA;
+		}
+	}
+
+	private void readWhilePossible(ChannelStateByteBuffer buffer) throws IOException {
+		while (haveMoreData() && buffer.isWritable()) {
+			if (remainingBytes <= 0) {
+				advanceOffset();
+			}
+			int bytesRead = serializer.readData(stream, buffer, remainingBytes);
+			remainingBytes -= bytesRead;
+		}
+	}
+
+	private boolean haveMoreData() {
+		return remainingBytes > 0 || !offsets.isEmpty();
+	}
+
+	@SuppressWarnings("ConstantConditions")
```

Comment:
`SuppressWarnings` is a compiler annotation so it can be ignored by humans.
Compiler warnings, on the other hand, are intended for humans and shouldn't be ignored. 
Therefore useless warnings produce clutter and unsupported `SuppressWarnings` not.

Reply:
My main point was that the specific warning is not a standard java compiler, but coming from IntelliJ and depends on your settings (I don't see this warning).
If I enable warning for unused suppression, I get a compiler warning for your suppression and I don't want to argue which settings are inherently better.
So I suggest to not suppress "ConstantConditions", but only standard suppressions like "unchecked".